aws/resource_aws_network_interface.go

```diff
@@ -444,3 +448,124 @@ func resourceAwsEniAttachmentHash(v interface{}) int {
 	buf.WriteString(fmt.Sprintf("%d-", m["device_index"].(int)))
 	return hashcode.String(buf.String())
 }
+
+func deleteNetworkInterface(conn *ec2.EC2, eniId string) error {
+	_, err := conn.DeleteNetworkInterface(&ec2.DeleteNetworkInterfaceInput{
+		NetworkInterfaceId: aws.String(eniId),
+	})
+
+	if isAWSErr(err, "InvalidNetworkInterfaceID.NotFound", "") {
+		return nil
+	}
+
+	if err != nil {
+		return fmt.Errorf("error deleting ENI (%s): %s", eniId, err)
+	}
+
+	return nil
+}
+
+func detachNetworkInterface(conn *ec2.EC2, eni *ec2.NetworkInterface, timeout time.Duration) error {
+	eniId := aws.StringValue(eni.NetworkInterfaceId)
+	if eni.Attachment == nil {
+		log.Printf("[DEBUG] ENI %s is already detached", eniId)
+		return nil
+	}
+
+	_, err := conn.DetachNetworkInterface(&ec2.DetachNetworkInterfaceInput{
+		AttachmentId: eni.Attachment.AttachmentId,
+		Force:        aws.Bool(true),
+	})
+
+	if isAWSErr(err, "InvalidAttachmentID.NotFound", "") {
+		return nil
+	}
+
+	if err != nil {
+		return fmt.Errorf("error detaching ENI (%s): %s", eniId, err)
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{
+			ec2.AttachmentStatusAttaching,
+			ec2.AttachmentStatusAttached,
+			ec2.AttachmentStatusDetaching,
+		},
+		Target: []string{
+			ec2.AttachmentStatusDetached,
+		},
+		Refresh:    networkInterfaceAttachmentStateRefresh(conn, eniId),
+		Timeout:    timeout,
+		Delay:      10 * time.Second,
+		MinTimeout: 5 * time.Second,
+	}
+
+	log.Printf("[DEBUG] Waiting for ENI (%s) to become detached", eniId)
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("error waiting for ENI (%s) to become detached: %s", eniId, err)
+	}
+
+	return nil
+}
+
+func networkInterfaceAttachmentStateRefresh(conn *ec2.EC2, eniId string) resource.StateRefreshFunc {
```

Comment:
BTW did you see that there was another networkInterfaceAttachmentRefreshFunc at line 207 above, in this very same file

Reply:
Thanks for pointing this out.
On deeper inspection there are a number of similar functions scattered throughout the code base.
My approach is usually to make the smallest sensible set of changes to address an issue raised as a bug or problem and then open (and fix, but at a slower pace) a second tech-debt issue to do the refactoring.

Reply:
I understand and this makes perfect sense. Thanks for digging deeper into this